lib/mongodb.js

```diff
@@ -1996,23 +1997,74 @@ function optimizedFindOrCreate(model, filter, data, options, callback) {
 }
 
 /**
- * Convert the decimal properties from string to decimal.
- * Only supported after mongodb 3.4
  *
- * @param {Object} data The data that might contain a decimal property
- * @param {Object} props The model property definitions
+ * @param {*} data Plain Data Object for the matching property definition(s)
+ * @param {*} modelCtorOrDef Model constructor or definition
+ * @param {*} setValue A callback function which takes a property value and
+ * definition to apply custom property coercion
  */
-function convertDecimalProps(data, props) {
-  if (debug.enabled) debug('convertDecimalProps props: ', util.inspect(props));
-  for (const p in props) {
-    const prop = props[p];
-    const isDecimal = data[p] && prop && prop.mongodb &&
-      prop.mongodb.dataType &&
-      prop.mongodb.dataType.toLowerCase() === 'decimal128';
-    if (isDecimal) {
-      data[p] = Decimal128.fromString(data[p]);
-      debug('convertDecimalProps decimal value: ', data[p]);
+function visitAllProperties(data, modelCtorOrDef, setValue) {
```

Comment:
Continuing discussion started earlier.

A function called `visitAllProperties` should not be modifying property values, it should only call the supplied `visiter` function for every property.

```suggestion
function visitAllProperties(data, modelCtorOrDef, visitor) {
```